sdks/go/test/integration/primitives/pardo.go

```diff
@@ -192,3 +196,90 @@ func emitPipelineOptions(_ []byte, emit func(string)) {
 	emit(fmt.Sprintf("%s: %s", "B", beam.PipelineOptions.Get("B")))
 	emit(fmt.Sprintf("%s: %s", "C", beam.PipelineOptions.Get("C")))
 }
+
+var CountInvokeBundleFinalizer atomic.Int32
+
+const (
+	BundleFinalizerStart   = 1
+	BundleFinalizerProcess = 2
+	BundleFinalizerFinish  = 4
+)
+
+// ParDoProcessElementBundleFinalizer creates a beam.Pipeline with a beam.ParDo0 that processes a DoFn with a
+// beam.BundleFinalization in its ProcessElement method.
+func ParDoProcessElementBundleFinalizer() *beam.Pipeline {
```

Comment:
Consider instead having the scope passed into this function *instead* of returning a Pipeline.

eg.

```
func ParDoProcessElementBundleFinalizer(s beam.Scope)  {
    imp := beam.Impulse(s)
	beam.ParDo0(s, &processElemBundleFinalizer{}, imp)
}
```

This cuts the boiler plate.

That makes it easier to integrated with the Prism Test suites, and thus ensure we get explicit code coverage over the prism you've written code.

eg. That's how we do the (increasingly mis-named) unimplemented_test.go file.

https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/runners/prism/internal/unimplemented_test.go#L74

So you could add the "Finalization" tests to the TestImplemented suite.
DO NOT refactor that test file to fix it, that's an orthogonal change that should be made separately.

Reply:
Naturally, the same for the other test cases in this file.